pkg/resources/tag_association.go

```diff
@@ -183,67 +166,168 @@ func CreateContextTagAssociation(ctx context.Context, d *schema.ResourceData, me
 			}
 		}
 	}
-	d.SetId(helpers.EncodeSnowflakeID(tid.DatabaseName(), tid.SchemaName(), tid.Name()))
+	d.SetId(helpers.EncodeResourceIdentifier(tagId.FullyQualifiedName(), tagValue, string(objectType)))
 	return ReadContextTagAssociation(ctx, d, meta)
 }
 
-func ReadContextTagAssociation(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	diags := diag.Diagnostics{}
+func ReadContextTagAssociation(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
 	client := meta.(*provider.Context).Client
+	tagValue := d.Get("tag_value").(string)
 
-	tid, ids, ot := TagIdentifierAndObjectIdentifier(d)
+	tagId, ids, objectType, err := TagIdentifierAndObjectIdentifier(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	var correctObjectIds []string
 	for _, oid := range ids {
-		tagValue, err := client.SystemFunctions.GetTag(ctx, tid, oid, ot)
+		objectTagValue, err := client.SystemFunctions.GetTag(ctx, tagId, oid, objectType)
 		if err != nil {
 			return diag.FromErr(err)
 		}
-		if err := d.Set("tag_value", tagValue); err != nil {
-			return diag.FromErr(err)
+		if objectTagValue != nil && *objectTagValue == tagValue {
+			correctObjectIds = append(correctObjectIds, oid.FullyQualifiedName())
 		}
 	}
-	return diags
+	if err := d.Set("object_identifiers", correctObjectIds); err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
 }
 
-func UpdateContextTagAssociation(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+func UpdateContextTagAssociation(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
 	client := meta.(*provider.Context).Client
-	tid, ids, ot := TagIdentifierAndObjectIdentifier(d)
-	for _, oid := range ids {
-		if d.HasChange("skip_validation") {
-			o, n := d.GetChange("skip_validation")
-			log.Printf("[DEBUG] skip_validation changed from %v to %v", o, n)
+	tagId, _, objectType, err := TagIdentifierAndObjectIdentifier(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if d.HasChange("tag_value") {
```

Comment:
in this two subsequent ifs we do exactly the same "setup" logic:
- we get the tag value
- we get the lists of ids (old, new, common)

Let's merge this logic:
1. If object is removed, unset the tag, no matter the value
2. If the object is added, set the tag, no matter if the value changed (we just use the new one in all cases)
3. If object is common, set only if the value changed

It covers all the cases (I guess) but is more straightforward

Reply:
Done.